BrowserView: Remove webview and initialize BrowserView.

vsvipul opened the pull request:
**What's this PR do?**
Implements BrowserView with similar functionality as webview. 
View is a class which we will use, it extends electron's BrowserView
class. ViewManager will be used to create, manage and delete views
in the main process. Removes all usage of webview and adds
an initial implementation of BrowserView.

**You have tested this PR on:**
  - [ ] Windows
  - [x] Linux/Ubuntu
  - [ ] macOS

priyank-p commented on app/main/view.ts:
```diff
@@ -0,0 +1,90 @@
+const shouldSilentWebview = ConfigUtil.getConfigItem('silent');
+
+class View extends BrowserView {
+	// TODO: Typescript - Create an interface for props here
```
Since you know what the props are going to be why this `TODO`? Just create the interface here.

vsvipul replied:
Done.

priyank-p commented on app/main/view.ts:
```diff
@@ -0,0 +1,95 @@
+import ConfigUtil = require('../renderer/js/utils/config-util');
+const shouldSilentWebview = ConfigUtil.getConfigItem('silent');
+
+export interface PropsType {
```
Name this `ViewProps`.

vsvipul replied:
Done.

priyank-p commented on app/main/view.ts:
```diff
@@ -0,0 +1,95 @@
+	badgeCount: number;
+	customCSS: boolean;
+
+	constructor(props: PropsType) {
```
Just a syntactic sugar you can do here: `constructor(public props: ViewProps)` and then remove `this.props = props;` below.

vsvipul replied:
Done.

vsvipul committed:
BrowserView: Add View wrapper Class for BrowserView.

View extends the BrowserView class in the main
process and provides and interface similar to
webview. Other functions can be added later
when required.

vsvipul committed:
BrowserView: Add ViewManager for managing multiple Views.

Adds ViewManager class which makes it easier
to create, manage and delete views. Also,
created some basic listeners which will be
required later.

vsvipul commented:
I think this is ready. @akashnimare @abhigyank . Should i wait for this to be merged or add subsequent commits on top of this PR ?

akashnimare commented:
>  add subsequent commits on top of this PR

Do this.

vsvipul committed:
BrowserView: Add BrowserView to index.ts.

vsvipul committed:
BrowserView: Remove webview props from tab.ts.

abhigyank commented:
I think you need to rebase with zulip:browserview branch otherwise the two extra commits from master would also come in the changes.

vsvipul commented:
> I think you need to rebase with zulip:browserview branch otherwise the two extra commits from master would also come in the changes.

I think it will be better to keep this branch updated with master as well, so that we don't face merge conflicts while merging into master later.

vsvipul committed:
BrowserView: Remove webview usage and initialize BrowserView.

Removes all instances of webview from main.ts and
adds an initial working instance of BrowserView.

vsvipul committed:
BrowserView: Add logic for badgeCount update.

Implements logic for updating badgeCount
to show message count in sidebar.

vsvipul committed:
BrowserView: Remove tooltips and use title instead.

Removes tooltips that we show for orgs and settings,
as these tooltips are no more visible over the
BrowserViews, so there is no use of keeping them.
Instead, uses title property of HTMLElements, which
can be displayed over BrowserViews.

vsvipul committed:
BrowserView: Add favicon-update event and fix view function calls.

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -0,0 +1,128 @@
+	}
+}
+
+export = new ViewManager();
```
Do `export default const ViewManager = new ViewManager();` or something else but not this `export =` syntax. We need to stop using this.

vsvipul replied:
```export default const ViewManager = new ViewManager();``` This doesn't seem to be working. Can you try this?

priyank-p replied:
So I made a mistake here, this could be either `export default new ViewManager()` or
```typescript
const viewManager = new ViewManager();
export default viewManager();

// then in index.ts
import ViewManager from './viewmanager';
```

I prefer the second one just because `export default new ViewManager();` looks awkward to me (but that just me).

akashnimare replied:
@vsvipul is this being taken care of?

abhigyank commented on app/renderer/js/components/server-tab.ts:
```diff
@@ -10,9 +10,8 @@ class ServerTab extends Tab {
 
 	template(): string {
 		return `<div class="tab" data-tab-id="${this.props.tabIndex}">
-					<div class="server-tooltip" style="display:none">${this.props.name}</div>
```
Any reason to remove this?
I see you removed all tooltips and moved to title. Any reason? Can we do the CSS for the title?

vsvipul replied:
No, we can't! Unfortunately we can't layer any HTML/CSS over BrowserViews, so this will not be possible.

abhigyank commented:
This is some really good work. One important observation that I have - Moving to BrowserView shifts almost everything to the main process. Most of the functions surrounding webview previously was in `renderer/main.js` since webview was in renderer. In this change, we have kept the structure same and in places where webview functions are called we are using ipc to move those calls to the main process - even for functions like create, reload, load, switch etc. Now that browserView resides in the main process, would it make sense for us move all these important browserView actions entirely to the main process from renderer/main instead of doing ipc calls?
I am not sure if ipc calls use much CPU or memory, but I think it would make sense to have the the basic initialisation actions moved to main process.

vsvipul commented:
> I am not sure if ipc calls use much CPU or memory, but I think it would make sense to have the the basic initialisation actions moved to main process.

@abhigyank  Yes, the initialization can be moved to main process, but switching tabs, and everything else, will still require ipc Calls.

vsvipul committed:
BrowserView: Destroy all views during reload-app.

Destroys all views to recreate them later
when app is reloaded. Also, set lastActiveIndex
to 0 when a tab is removed.

vsvipul committed:
BrowserView: Improvements to loading indicator.

priyank-p commented on app/main/view.ts:
```diff
@@ -0,0 +1,93 @@
+'use strict';
```
Minor change, but the commit message for this commit could be:
```
BrowserView: Add View wrapper class for webview.

View class extends the BrowserView class and provides interface
similar to of webview. Other functions can be later when required.
```

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -0,0 +1,128 @@
+'use strict';
```
The commit message for this commit should be:
```
BrowserView: Add ViewManger for managing multiple View classes.

The ViewManager class makes it easier to manager multiple Views.
```

because the `ViewManager` class is not a wrapper. The summary can be extended.

priyank-p replied:
Can you also add more details in commit message how does it handle multiple view
exactly in more detail, like resize etc.?

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -0,0 +1,128 @@
+			this.destroyAll();
+		});
+
+		ipcMain.on('forward-message-view', (e: Event, name: string, ...params: any[]) => {
```
The event name should be `forward-view-message`?

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	registerIpcs(): void {
+		ipcMain.on('view-create', (e: Event, props: ViewProps) => {
```
This sounds, and reads for that matter, better as `create-view`, `select-view` and `destroy-view`?

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -0,0 +1,128 @@
+			this.destroy(index);
+		});
+
+		ipcMain.on('view-destroy-all', () => {
```
I guess since I am posting review about event names, `destroy-all-views` here?

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	create(props: ViewProps): void {
+		if (this.views[props.index]) {
```
I think I mentioned this before but, the allowing a programmer to pass and track IDs to pass into this `index` field here is very brittle; it seems to call for trouble. Can we make this auto generated and make `readonly` property?

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	fixBounds(): void {
+		const SIDEBAR_WIDTH = 54;
```
I'd add a comment here mentioning that any updates here should be reflected to css and other way around.

priyank-p reviewed (commented):
@vsvipul I reviewed the first two commits and left minor comments. I don't like the passing index programmatically and having to manage it; I do rather the code handle that. Let's come up with a better model here, I guess we can come up with a better model if you describe what exactly it's needed for.

vsvipul commented:
> @vsvipul I reviewed the first two commits and left minor comments. I don't like the passing index programmatically and having to manage it; I do rather the code handle that. Let's come up with a better model here, I guess we can come up with a better model if you describe what exactly it's needed for.

Index is needed to create a view with that particular index so that we can easily remove and change to that view using only the index later, nothing else is required. What better model do you propose?

vsvipul committed:
BrowserView: Improve ipc names.

priyank-p commented:
@vsvipul OK, here is what we can do, we can have a internal variable in `view.t` say `viewIndex` that is set to 0 at first. Then when we create a view in `constructor` we can do `this.index = viewIndex;` and increment it `viewIndex++`. That way we can access the index and do whatever we want like remove, select but we just don't have to pass it in.

priyank-p commented on app/main/index.ts:
```diff
@@ -262,6 +263,10 @@ app.on('ready', () => {
 		page.send('toggle-autohide-menubar', showMenubar, true);
 	});
 
+	ipcMain.on('toggle-sidebar', () => {
```
Do we need to do similar thing for app window resize (or is it done already)?

vsvipul replied:
```
view.setAutoResize({ width: true, height: true });
```
This line in fixBounds() ensured automatic resizing with window.

priyank-p replied:
Great!

priyank-p commented on app/main/menu.ts:
```diff
@@ -89,7 +89,7 @@ class AppMenu {
 			click(_item: any, focusedWindow: any) {
 				if (focusedWindow) {
-					focusedWindow.webContents.toggleDevTools();
+					focusedWindow.webContents.openDevTools({mode: 'undocked'});
```
How do we handle the toggle behavior here, since this just seems to open devtools.

vsvipul replied:
Yes, it opens DevTools in a new window. We can't toggle it. This was the case earlier for Opening devtools for active tab. But, not we are showing devtools for active tab in the same window and for zulip app in a new window, because zulip app devtools are overlapped by the browser views.

priyank-p replied:
So, there is no way to toggle behavior like we had before?

vsvipul replied:
No, we can't toggle something that opens in a new window. We coudn't toggle the tools for active tab earlier, now we can't toggle for zulip app, and vice versa.

priyank-p commented on app/main/view.ts:
```diff
@@ -17,7 +17,7 @@ export class View extends BrowserView {
 	index: number;
 	url: string;
 	zoomFactor: number;
-	loading: true;
```
Wow, I am surprised typescript didn't complain here.

vsvipul replied:
Removed loading completely from here and moved it to main.ts

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -41,12 +41,9 @@ class ViewManager {
-			}
+			// Type checking requires spread elements to match up with a rest parameter.
+			// So, using a workaround here.
+			(this.views[this.selectedIndex] as any)[name as keyof View](...params);
```
I wonder if `(this.views[this.selectedIndex] as View)` work here?

vsvipul replied:
No, can't. Getting this error-
```
Cannot invoke an expression whose type lacks a call signature. Type 'string | number | boolean | ViewProps | WebContents | (() => boolean) | ((event: string, listener: Function) => View) | ((event: string, listener: Function) => View) | ((event: string, listener: Function) => View) | ... 27 more ... | (() => string[])' has no compatible call signatures.ts(2349)
```

priyank-p replied:
OK, so I think I know why this throws error. The correct way to write this would be:
```typescript
type ViewCallable = (...params: any[]) -> any;
(this.views[this.selectedIndex][name as keyof View] as ViewCallable)(...params);
```

priyank-p replied:
Since we do this in couple of places I do split this off into it's own function and also check if it a function before calling it too. So something like:
```typescript
callViewFunction(index: number, name: keyof View, ...params; any[]) {
    const viewFunction = this.views[index][name];
    if (typeof viewFunction === 'function') {
        type ViewCallable = (...args: any[]) -> any;
        (viewFunction as ViewCallable)(...params);
    } else {
         // error reporting here
    }
}
```

vsvipul replied:
I have made a separate function in the last commit, but couldn't make this work. Get webcontents not defined error when i use this. See if you can make it work.

priyank-p commented on app/main/view.ts:
```diff
@@ -45,6 +45,34 @@ export class View extends BrowserView {
+			if (isSettingsPage) {
+				return;
+			}
+			this.canGoBackButton();
```
Is this function call for showing the go back button if we can, or to be more specific enable the button. If this is the case I wonder if `maybeEnableGoBackButton` would be a better name.

vsvipul replied:
I have used the same name as in webview.ts. Should i change this ?

priyank-p replied:
Sure, `maybeEnableGoBackButton` sounds better.

vsvipul replied:
Ok i'll change the name here.

abhigyank replied:
`maybeEnableGoBackButton` sounds like a hackish function name to me.

priyank-p commented on app/main/viewmanager.ts:
```diff
@@ -67,6 +64,7 @@ class ViewManager {
 			return;
 		}
 		this.selectedIndex = index;
+		mainWindow.setBrowserView(null);
```
What is the story behind this change?

vsvipul replied:
Unsetting the old view by doing this and then setting the new view ensures that there are no glitches while switching. Can be removed too, but i guess this is better.

priyank-p replied:
OK, I'd add a comment stating that here.

vsvipul replied:
Did you add anything earlier too? I did a force push recently.

priyank-p replied:
What do you mean by "Did you add anything"?

priyank-p reviewed (commented):
Left some more comments, I like how we get to change `tab.webview.props.name` -> `tabs.props.name`.

vsvipul committed:
BrowserView: Add logic to handle external links.

vsvipul committed:
BrowserView: Rename canGoBackButton to maybeEnableGoBackButton.

akashnimare commented on app/main/index.ts:
```diff
@@ -261,6 +264,10 @@ app.on('ready', () => {
 		page.send('toggle-autohide-menubar', showMenubar, true);
 	});
 
+	ipcMain.on('fix-bounds', () => {
```
Needs to document this.

vsvipul replied:
Will do after the testing and bug fixing is done.

vsvipul replied:
DOne.

akashnimare commented on app/main/menu.ts:
```diff
@@ -89,7 +89,7 @@ class AppMenu {
 			click(_item: any, focusedWindow: any) {
 				if (focusedWindow) {
-					focusedWindow.webContents.toggleDevTools();
+					focusedWindow.webContents.openDevTools({mode: 'undocked'});
```
This `undocked` thing open the devtools, right?

vsvipul replied:
Yes, it opens them in a new window, as we can't show them inside the window now.

akashnimare commented on app/main/view.ts:
```diff
@@ -0,0 +1,174 @@
+import { BrowserView, BrowserWindow, app } from 'electron';
+
+import ConfigUtil = require('../renderer/js/utils/config-util');
+const shouldSilentWebview = ConfigUtil.getConfigItem('silent');
```
Should we set a backup value for this? In case the json gets corrupted? i.e.
```
const shouldSilentWebview = ConfigUtil.getConfigItem('silent') || false;
```

vsvipul replied:
I picked it up from webview.ts . Is this necessary? If yes, then i'll change it.

akashnimare replied:
Leave it for now then.

akashnimare commented on app/main/view.ts:
```diff
@@ -0,0 +1,174 @@
+}
+
+export class View extends BrowserView {
+	index: number;
```
Should we make an interface for this also?

vsvipul replied:
Not required since we haven't used it anywhere.

vsvipul committed:
BrowserView: Implement userAgent and updateBadge onload.

Add logic to set user agent for each browserview.
Also, the badgeCount was not shown initially,
on opening the app. Adds that too.

vsvipul committed:
BrowserView: Add logic to show network error page.

vsvipul committed:
BrowserView: Add logic for custom CSS.

vsvipul committed:
BrowserView: Call fixBounds before setting view.

vsvipul committed:
BrowserView: Update deps.

vsvipul committed:
gitignore: Add .vscode to gitignore.

vsvipul committed:
BrowserView: Show views only when dom content loaded.

Adds a refresh function which refreshes view
every 200 ms, and shows it only if the view
DOM content is loaded. Also, gets rid of buggy
browser view switching which we were facing earlier.

vsvipul commented:
@akashnimare @abhigyank @priyank-p @kanishk98 Can you pull and test this? I think this PR is complete if the last commit is good to go. (Still removing webview file is left).

vsvipul committed:
BrowserView: Rename webview instances to view and handle tray toggle.

abhigyank commented on app/main/index.ts:
```diff
@@ -279,7 +285,7 @@ app.on('ready', () => {
 		const activeTab = props.tabs[props.activeTabIndex];
 		if (activeTab) {
-			mainWindow.setTitle(`Zulip - ${activeTab.webview.props.name}`);
+			mainWindow.setTitle(`Zulip - ${activeTab.props.name}`);
```
It seems on the BrowserView, it doesn't auto decode encoded text (in case of abhigyank.zulipdev.org) so we see escaped HTML on the title. For temporary fix we can decode the title here with a comment and create an issue upstream to decode the url. It used to work webview so should work here as well.

vsvipul replied:
Fixed this in the latest commit. We weren't decoding it when we were pushing it to tabs.

abhigyank commented:
Bug - If you resize the window, all other servers than the active one gets reloaded. Same is the case with hiding sidebar or menu bar. If you come back to the active server after switching that reloads too. Though I don't feel that its a full reload or something cause its quite fast. 

Rest tab switching and everything are working fine and smooth. I couldn't find any bugs. Haven't gone through the implementation and code yet but seems to be working okay. @priyank-p Maybe you can give it a try if you can find something broken.

vsvipul committed:
BrowserView: Add decoded server name when pushing to tabs.

vsvipul commented:
> Bug - If you resize the window, all other servers than the active one gets reloaded. Same is the case with hiding sidebar or menu bar. If you come back to the active server after switching that reloads too. Though I don't feel that its a full reload or something cause its quite fast.

This is known. Still looking into what is causing this. Couldn't find what is causing this weird behavior @priyank-p  .

vsvipul commented:
> Bug - If you resize the window, all other servers than the active one gets reloaded. Same is the case with hiding sidebar or menu bar. If you come back to the active server after switching that reloads too. Though I don't feel that its a full reload or something cause its quite fast.

This has something to do with our preload script. Try logging out of all orgs and do the same. It will be fixed. It isn't something related to BrowserView. It is something in our own preload script (specifically spellchecker i think, which is causing this problem). What should i do about it?

vsvipul committed:
BrowserView: Make separate callViewFunction.

vsvipul committed:
BrowserView: Fix focus switching.

akashnimare commented:
> This has something to do with our preload script. Try logging out of all orgs and do the same. It will be fixed. It isn't something related to BrowserView. It is something in our own preload script (specifically spellchecker i think, which is causing this problem). What should i do about it?

@vsvipul can you reproduce this on master?

vsvipul commented:
> @vsvipul can you reproduce this on master?

No, i can't reproduce this on master. What i meant to say was that, it has something to with spellchecker and BrowserView compatibility with each other that seems to causing this. I've been trying to fix this from the past week, but couldn't fix it. If someone can please help, it'd be great!

kanishk98 commented:
> What i meant to say was that, it has something to with spellchecker and BrowserView compatibility with each other that seems to causing this.

@vsvipul what makes you say that? I commented out the spellchecker related actions in `preload.ts` but that didn't solve the bug. 
I think it's probably something else - I'll let you know if I find something. :)

vsvipul committed:
Remove unnecessary translation-util.js file.

vsvipul commented:
> @vsvipul what makes you say that? I commented out the spellchecker related actions in `preload.ts` but that didn't solve the bug.
> I think it's probably something else - I'll let you know if I find something. :)

Sorry, i just tested it again and i guess i was wrong! @akashnimare  Is this required to be fixed before merging this PR? I'll be adding the documentation today.

vsvipul committed:
BrowserView: Remove show dom content on load.

Removes the heavy setInterval which we were
using earlier to prevent heavy CPU Usage.
The main reason we were using a setInterval
earlier was because view switching was buggy.
This commit fixes that.

akashnimare commented on app/renderer/js/components/handle-external-link.ts:
```diff
@@ -22,8 +20,6 @@ function handleExternalLink(this: any, event: any): void {
 	} = LinkUtil.isInternal(domainPrefix, url);
 
 	if (isWhiteListURL) {
-		event.preventDefault();
```
Can't remember exactly why we added this but I think it has something to do with the race condition where the link was being opened in the browser + the desktop app. Please cross-check it.

vsvipul replied:
For what type of link do i need to test this?

akashnimare replied:
Try to test the following links -
* Internal url - Click on internal zulip link like send a message to yourself with something like https://chat.zulip.org/#narrow/stream/112-desktop-issues and click the same.
* Click on an attachment link.
* Some external URL.

akashnimare replied:
@vsvipul Is this being tested?

akashnimare commented on app/renderer/js/components/handle-external-link.ts:
```diff
@@ -8,10 +8,8 @@ const { shell, app } = remote;
 
-// TODO: TypeScript - Figure out a way to pass correct type here.
-function handleExternalLink(this: any, event: any): void {
-	const { url } = event;
```
I'm guessing it's a legacy code which we forgot to remove?

vsvipul replied:
Yes, typescript detected that here, and it is passed correctly now.

vsvipul committed:
BrowserView: Add documentation and fix fullScreen.

akashnimare commented on app/main/viewmanager.ts:
```diff
@@ -0,0 +1,161 @@
+			this.callViewFunction(this.selectedIndex, name, ...params);
+		});
+
+		ipcMain.on('call-specific-view-function', (e: Event, index: number, name: string, ...params: any[]) => {
```
Please document this newly added IPCs -
> forward-message-all
> call-view-function
> call-specific-view-function
> forward-view-message

@priyank-p happy with the naming of ipc functions?

vsvipul replied:
Documented these.

abhigyank replied:
I think we can rename `forward-view-message` to `forward-message-view` and `forward-message-all` to `forward-message-all-views`. Similarly the `forwardMessageAll` to `forwardMessageAllViews`. @akashnimare thoughts?

akashnimare commented on app/renderer/js/components/tab.ts:
```diff
@@ -27,17 +23,14 @@ class Tab extends BaseComponent {
-		this.webview.hide();
 	}
 
 	destroy(): void {
```
Are we using these listeners anywhere?
> activate
> deactivate
> destroy

If yes, please document this.

vsvipul replied:
Done.

akashnimare commented on app/main/index.ts:
```diff
@@ -108,10 +110,12 @@ function createMainWindow(): Electron.BrowserWindow {
 	win.setTitle('Zulip');
 
 	win.on('enter-full-screen', () => {
+		ViewManager.fixBounds();
```
Can you document the bound thing here and what it do etc?

vsvipul replied:
I have documented this here. https://github.com/zulip/zulip-desktop/blob/e4f025166f882ffa9bd6fcde73fe19d4d7572d76/app/main/viewmanager.ts#L106

akashnimare commented:
This is getting pretty close to being merged. @vsvipul please take care of the conflicts. @abhigyank, @kanishk98, @priyank-p would be great if you guys can test the PR from your side.

vsvipul committed:
Add integer check for messageCountAll

akashnimare commented:
We have to look for the potential issues in electron-builder which could affect our app since we are using v21 of electron-builder -
https://github.com/electron-userland/electron-builder/issues/4185

akashnimare commented:
@vsvipul  I get this after running `npm start` -

https://cl.ly/08d596abfc9a

Tried reinstalling node modules and `npm run clean-ts-files` but no luck.

vsvipul commented:
@akashnimare This is working fine for me. Wonder what's causing the problem for you. @kanishk98  @abhigyank  @priyank-p  can you guys test once again as well? Let's wrap this up.

akashnimare commented:
@vsvipul I'm still stuck to that page. For me, even the setting page is not loading. No error in console.

![image](https://user-images.githubusercontent.com/2263909/65038503-b7cf7580-d96d-11e9-837e-e8b7642c4b44.png)

akashnimare commented:
@kanishk98 can you check if this PR works for you on macOS? This works for me on Linux but something is wrong on macOS.

kanishk98 commented:
Yeah, it doesn't work on macOS for me either. (screenshot attached below)
I remember it working on Windows last time I tried it, though. @vsvipul are we using Windows/Linux specific APIs somewhere?

![Screenshot 2019-09-18 at 5 36 29 PM](https://user-images.githubusercontent.com/24617297/65147219-507ff700-da3b-11e9-820e-004e7a60ea0d.png)

akashnimare commented:
In that case, we should check with electron v5 and see if they have made any changes.

akashnimare commented:
@kanishk98 let's try to find out which commit broke the app on macOS.

kanishk98 commented:
@akashnimare I'm doing that right now. I don't think it's something to do with Electron, though - this seems to stem from a bug in `activateTab()`. I'm trying to fix it right now, will text you and Vipul on czo if I find something.

akashnimare commented:
@kanishk98 any luck in debugging?

kanishk98 commented:
> any luck in debugging?

Yeah, the app breaks at commit 9c4a3de835d0e7f20948e07982d980f03bb2dfc7. I'm currently working to see where the Mac-specific problem is.

kanishk98 commented:
Ah okay, the solution is almost too simple:

In `viewmanager.select()`, we need to reverse the order of function calls from 

```ts
this.fixBounds();
mainWindow.setBrowserView(view);
```

to

```ts
mainWindow.setBrowserView(view);
this.fixBounds();
```

@vsvipul can you make the changes? I just tried this out on a Mac - it should solve our problem.

akashnimare commented:
@kanishk98 are you sure this works for you? Can you post a gif here?

kanishk98 commented:
> Can you post a gif here?

I can't post a gif right now (it's not my Mac that I tested this on), but you can fetch my branch [browserview](https://github.com/kanishk98/zulip-desktop/tree/browserview) onto your system and try it out. Do let me know if it doesn't work.

akashnimare commented:
@vsvipul can you fix the merge conflicts?

vsvipul commented:
@akashnimare working on it right now. Sorry for the late response. Was a little busy this week.

vsvipul commented:
I am having some difficulty porting the new notification option to browser view.

kanishk98 commented:
> I am having some difficulty porting the new notification option to browser view.

I think I can help with that. Can you tell me what the issue is?

vsvipul commented:
@kanishk98 How do i set the enabled state in the rendered process for that setting because the current URL in browser view is in the main procesS?

kanishk98 commented:
I can think of the following methods:

1. Use a synchronous IPC call (probably the simplest, but I'd avoid this). 
2. Save a `loggedIn` flag in `domain.json` for all orgs that you can read when deciding whether `enabled` is true or not. Said flag is updated when the URL of the view changes. 
3. I'm a little doubtful about this one, but we can also maintain an instance variable in `ViewManager` that maps server URLs to `loggedIn` status. Updated in the same way as (2). 

Since I'm also working on this right now, I'll try what works on my end and let you know. Text me on czo if you have any further questions.

akashnimare commented:
@kanishk98, @vsvipul awesome. Let's try to merge this asap since it's a blocker for other PRs.

vsvipul commented:
@akashnimare  Apparently, rebasing it automatically messes up my whole branch. The title property which I added gets removed, function names get changed due to which many ipc Calls stop working. I'll have to do this commit by commit myself. So, it is going to take some time.

akashnimare commented:
@vsvipul maybe reset the branch locally and git pull + rebase and fix merge conflicts? There are three files mainly where the conflicts thing happening.

vsvipul commented:
@akashnimare Yeah, i did exactly that, but that still messes up a lot of stuff. I am doing it but will take some time.

akashnimare commented:
@vsvipul let me know if you need help. Gotta merge this asap since a lot of other PRs are still pending.

vsvipul commented:
@akashnimare As I am currently a little busy with academic work, I have given access to my branch to @kanishk98 a few days back. He is working on rebasing and migrating the notification option he added earlier.

akashnimare commented:
Cool, thanks for the update!

vsvipul commented:
Closing this in favor of #831 .

vsvipul commented:
Yes @andersk . Thanks. Updated.
